### DISCUSSION TITLE
Signup for Coverty Static Analysis

### DISCUSSION THREAD

**Discussion by ryao** [NVIDIA/open-gpu-kernel-modules#241] [Ideas]
https://scan.coverity.com/

This is a rather nice tool that the Linux kernel and other projects use to find certain classes of bugs before users report them. It makes it easy for potential contributors to find things that need to be fixed so that they can submit patches and should help increase community involvement in the kernel driver’s development. It also has GitHub integration.

Please sign up for it.

**Comment by PAR2020** [comment]
Thanks Richard, for your suggestion. In addition to the other tools (lint, CI), we have also been discussing for use here, while blending into the tools used internally. Will keep you tuned into where we land.

**Reply by ryao** [reply]
I just noticed that Nvidia appears to already use coverity, but not in the public eye:

https://github.com/NVIDIA/open-gpu-kernel-modules/search?q=coverity&type=code

There is no reason to keep this private, since anyone can fork the repository and upload builds to coverity,

**Reply by mtijanic** [reply]
Hi @ryao . Yes, we do use coverity internally, but we use a self-hosted instance that is only accessible through VPN, so making those builds public would be prohibitively complicated, sorry.

Another thing to note is that while we do run coverity on the CPU/GSP driver model, we currently use the proprietary codebase for this, and not the published open code. This means that it would likely find many defects in the open code that exist only because parts of the code are not published yet - this is the cause of so many "unused variable" instances in the codebase.
In r515 we explicitly disabled much of non-functional code that relates to GeForce and Workstation features on GSP. We are actively working on enabling those, and that will fix many of these coverity defects as a side-effect. In my own opinion, it does not make sense to enable coverity scans here until we get all (or most) of the code in place first.

**Reply by throwaway1037** [reply]
> many defects in the open code that exist only because parts of the code are not published yet

As this driver is just a fraction of the proprietary driver, is NVIDIA planning on publishing the rest of the proprietary driver as free/libre software?
The way you phrased it by saying "yet" leads me to believe there is still hope this will happen.

**Reply by ryao** [reply]
They are writing new code to fill in those bits and should be referring to that.

**Reply by throwaway1037** [reply]
Why not just liberate the existing proprietary code instead of writing new code?
Are they trying to hide what's in the proprietary code?
Why not release, let alone liberate, PMU firmware?

Thia all just comes off as a deliberate attempt to stop Nouveau writing power management features and becoming a competitive driver.

**Reply by ryao** [reply]
I do not think this is the correct forum to complain about what Nvidia decides to do with their older source code.

**Reply by YusufKhan-gamedev** [reply]
> Why not just liberate the existing proprietary code instead of writing new code?
> Are they trying to hide what's in the proprietary code?
> Why not release, let alone liberate, PMU firmware?

Non-free dependencies, non-renegotiable contracts that include a NDA on code written, etc.
